R/utils-get_code_dependency.R

```diff
@@ -26,22 +26,23 @@
 #'
 #' @keywords internal
 get_code_dependency <- function(code, names, check_names = TRUE) {
-  checkmate::assert_character(code)
+  checkmate::assert_list(code, "character")
   checkmate::assert_character(names, any.missing = FALSE)
 
-  if (identical(code, character(0)) || identical(trimws(code), "")) {
+  if (length(code) == 0) {
     return(code)
   }
 
   # If code is bound in curly brackets, remove them.
+  # TODO: rethink if this is still needed when code is divided by calls?
   tcode <- trimws(code)
   if (any(grepl("^\\{.*\\}$", tcode))) {
-    code <- sub("^\\{(.*)\\}$", "\\1", tcode)
+    tcode <- sub("^\\{(.*)\\}$", "\\1", tcode)
   }
 
+  parsed_code <- parse(text = tcode, keep.source = TRUE)
 
-  code <- parse(text = code, keep.source = TRUE)
-  pd <- utils::getParseData(code)
+  pd <- utils::getParseData(parsed_code)
```

Comment:
Watch out @m7pr . Original `pd` has different rows order than one returned from `extract_calls(pd)`. `extract_calls` parsing `pd` recursively and returns a list of calls where rows are topologically ordered. Now we don't need `extract_calls` because the whole code is split into single codes and we just need to `parse` each of them. Problem is that if we just parse each code element we would get original `pd` which will break `extract_occurrence`. We either need to topologically sort pd or fix `extract_occurrence`. Quick example.
Please look at the difference below, assign operator is after assigned symbol in the original pd:

```
pd <- getParseData(parse(text ="x <- y", keep.source = TRUE))
teal.code:::extract_calls(pd)
#   line1 col1 line2 col2 id parent       token terminal text
# 7     1    1     1    6  7      0        expr    FALSE     
# 1     1    1     1    1  1      3      SYMBOL     TRUE    x
# 3     1    1     1    1  3      7        expr    FALSE     
# 2     1    3     1    4  2      7 LEFT_ASSIGN     TRUE   <-
# 4     1    6     1    6  4      6      SYMBOL     TRUE    y
# 6     1    6     1    6  6      7        expr    FALSE     
# > teal.code:::extract_calls(pd)
# [[1]]
#         token text id parent
# 7        expr       7      0
# 3        expr       3      7
# 2 LEFT_ASSIGN   <-  2      7
# 6        expr       6      7
# 1      SYMBOL    x  1      3
# 4      SYMBOL    y  4      6
```

Reply:
@gogonzo did you try running `extract_occurrence` on it?

```r
> teal.code:::extract_occurrence(teal.code:::extract_calls(pd))
[[1]]
[1] "x"  "<-" "y" 
```

Reply:
We can discuss this once `get_code_dependency` works again. I put this comment based on the assumption that you're going to put `occurrence` and `side_effect` as a attribute of the `code` in `eval_code`. Let's ignore this comment at this moment, it might be relevant when you propose change of the `get_code_dependency`/`get_code`